anymail/webhooks/postal.py

```diff
@@ -0,0 +1,175 @@
+    webhook_key = None
+
+    def __init__(self, **kwargs):
+        self.webhook_key = get_anymail_setting('webhook_key', esp_name=self.esp_name, kwargs=kwargs, allow_bare=True)
+
+        super().__init__(**kwargs)
+
+    def validate_request(self, request):
+        try:
+            signature = request.META["HTTP_X_POSTAL_SIGNATURE"]
+        except KeyError:
+            raise AnymailWebhookValidationFailure("X-Postal-Signature header missing from webhook")
+
+        public_key = serialization.load_pem_public_key(
+            ('-----BEGIN PUBLIC KEY-----\n' + self.webhook_key + '\n-----END PUBLIC KEY-----').encode(),
+            backend=default_backend()
+        )
+
+        try:
+            public_key.verify(
+                b64decode(signature),
+                request.body,
+                padding.PKCS1v15(),
+                hashes.SHA1()
+            )
+        except InvalidSignature:
+            raise AnymailWebhookValidationFailure(
+                "Postal webhook called with incorrect signature")
+
+
+class PostalTrackingWebhookView(PostalBaseWebhookView):
+    """Handler for Postal message, engagement, and generation event webhooks"""
+
+    signal = tracking
+
+    def parse_events(self, request):
+        esp_event = json.loads(request.body.decode("utf-8"))
+
+        raw_timestamp = esp_event.get("timestamp")
+        timestamp = (
+            datetime.fromtimestamp(int(raw_timestamp), tz=utc)
+            if raw_timestamp
+            else None
+        )
+
+        payload = esp_event.get("payload", {})
+
+        status_types = {
+            "Sent": EventType.DELIVERED,
+            "SoftFail": EventType.DEFERRED,
+            "HardFail": EventType.FAILED,
+            "Held": EventType.QUEUED,
+        }
+
+        def get_event_type(event):
+            if "status" in event:
+                return status_types.get(event["status"], EventType.UNKNOWN)
+
+            if "bounce" in event:
+                return EventType.BOUNCED
+
+            if "url" in event:
+                return EventType.CLICKED
+
+            return EventType.UNKNOWN
+
+        event_type = get_event_type(payload)
+
+        description = payload.get("details")
+        mta_response = payload.get("output")
+
+        # extract message-related fields
+        message = payload.get("message") or payload.get("original_message", {})
+        message_id = message.get("message_id")
+        tag = message.get("tag")
+        recipient = None
+        message_to = message.get("to")
+        if message_to is not None:
+            try:
+                recipient = parse_single_address(message_to).addr_spec
+            except AnymailInvalidAddress:
+                pass
+
+        if message.get("direction") == "incoming":
+            # This is a tracking event about inbound email, so it's not interesting to anymail.
+            # We can safely ignore this.
+            return []
```

Comment:
For other ESPs, Anymail raises a ConfigurationError ([example](https://github.com/anymail/django-anymail/blob/v8.1/anymail/webhooks/postmark.py#L79-L82)) in this case, because swapping webhook urls is a pretty common configuration mistake. (Assuming Postal lets you set up different webhooks for different event types; I can't tell from their docs.)

Reply:
done! 

I also tried to clarify the purpose of the code above, does it make sense to you?

```
            # Let's ignore tracking events about an inbound emails.
            # This happens when an inbound email could not be forwarded.
            # The email didn't originate from Anymail, so the user can't do much about it.
            # It is part of normal Postal operation, not a configuration error.
```